Playwright: add screenshot capability

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

Implements ability to save screenshots for failing tests.

**Root Hooks**
Use root hooks to implement screenshot capture for failing tests.

**calypso-e2e package**
Implement helper functions for common operations (eg. screenshot directory)
Implement function called by the root level hook within the `calypso-e2e/src/hooks` directory.

#### Testing instructions

TeamCity: check if test fails and artifact shows screenshots.
[example test run where it failed as expected](https://teamcity.a8c.com/buildConfiguration/calypso_RunCalypsoPlaywrightE2eTests/6170727?buildTab=artifacts)

Run locally and verify the screenshots are saved at the expected locations:
- magellan: temp/build-xxxx directory
- mocha: packages/calypso-e2e/dist/screenshots

Partially implements #51082.
Parent: #51285
Child: #52678

worldomonation commented:
Looks like screenshot path is still a problem.

scinos commented on test/e2e/lib/mocha-hooks.js:
```diff
@@ -51,8 +52,51 @@ afterEach( function () {
-// Take Screenshot
+// Take Screenshot for Playwright.
 afterEach( async function () {
+	if ( ! global.isPlaywright ) {
```
(For a future PR, noting it here so we don't forget)

This can use some refactoring. I think we can have a `mocha/hooks/` with many individual hooks (eg: `mocha/hook/playwright-screenshot.js`), an index that imports some hooks (eg: `mocha/playwright-hooks.js`), and specify it inside the `magellan-*.json` config file with `mocha_args: '--file mocha/playwright-hooks.js'`

worldomonation replied:
Interesting, I had not known about this ability and quick search with relevant keywords don't turn up any results. It might entrench our reliance on Magellan for a bit longer, but that would help me remove a bunch of unclean if/else statements from the hook code.

scinos commented on test/e2e/lib/mocha-hooks.js:
```diff
@@ -51,8 +52,51 @@ afterEach( function () {
+	const screenSize = browserManager.targetScreenSize().toUpperCase();
+	const locale = browserManager.targetLocale().toUpperCase();
+	const date = new Date().getTime().toString();
+	const fileName = `FAILED-${ locale }-${ screenSize }-${ shortTestFileName }-${ date }`;
```
Shouldn't `FAILED` be dynamic and depend on `this.currentTest.state`? Otherwise, when `saveAllScreenshots:true` all screenshots will be `FAILED-*.png`, even for passing tests.

scinos commented on test/e2e/lib/mocha-hooks/playwright-hooks.ts:
```diff
@@ -4,22 +4,68 @@
+	// Page represents the target page to be captured.
+	// Reference to Page object is set during beforeEach hook in the describe block
+	// in the currentTest's context.
+	return await this.currentTest?.ctx?.page.screenshot( { path: screenshotPath } );
```
No need to `return` anything, afaik Mocha will ignore the return value of hooks.

worldomonation committed:
Implement Playwright screenshot capture on a new branch.

- move the MediaHelper file inside the calypso-e2e package.
- write a new afterEach hook for screenshot capture.

Purposely introduce a failure to test TeamCity screenshot configuration.

Remove intentional failure from test spec.

Rebase working branch onto updated parent.

noahtallen commented on test/e2e/lib/hooks-playwright/hooks.ts:
```diff
@@ -14,6 +15,48 @@ before( 'Launch browser instance', async function () {
+	// Page represents the target page to be captured.
+	// Reference to Page object is set during beforeEach hook in the describe block
+	// in the currentTest's context.
+	return await this.currentTest.page.screenshot( { path: screenshotPath } );
```
Related to my comment in the other pr, I think it would be good if we don't have to rely on people to remember to set the current page for every test suite.

worldomonation committed:
Move screenshot hook function within the @automattic/calypso-e2e package.

worldomonation committed:
Move the saveScreenshot hook to use the new root-level hook approach suggested by mocha 8.0.0.

worldomonation committed:
Add @types/mocha to @automattic/calypso-e2e and yarn lock file.

noahtallen commented on packages/calypso-e2e/src/index.ts:
```diff
@@ -2,3 +2,5 @@
  * Internal dependencies
  */
 export * as BrowserManager from './browser-manager';
+export * as MediaHelper from './media-helper';
```
IMO, we could keep these helpers internal since they aren't used outside of the package. (unless they are :P)

worldomonation replied:
Good point, I thought I _had to_ export in order for the package to work. Will make it internal.

noahtallen replied:
you should be able to import the media helper file directly from the screenshots hooks file

worldomonation replied:
Removed the export in the current revision.

noahtallen reviewed (approved):
Looks like screenshots are working, so this makes sense to me. I assume the "blank" screen showing up in them is just because the screen isn't loading anything, not a problem with the screenshots?

worldomonation commented:
> Looks like screenshots are working, so this makes sense to me. I assume the "blank" screen showing up in them is just because the screen isn't loading anything, not a problem with the screenshots?

Yep, it's literally taking a screenshot of a blank page. I should probably have tried for something more creative :)

As always, I really appreciate that you take time out of your day to review my PRs. Hopefully I can put forth more complete PRs going forward.

scinos committed:
Try bail instead of mocha-steps

scinos committed:
Fix test

worldomonation commented:
I rebased this branch on top of @scinos ' `ci/e2e/try-bail` branch and everything ran perfectly fine, including screenshot capture. Yay!

This was the output after I included the `--bail` option:
```
  Main Suite 1 @parallel
    Subsuite 1-1
      ✓ Can see the log in page (533ms)
    Subsuite 1-2
      ✓ Should also pass (384ms)

  Main Suite 2 @parallel
    Subsuite 2-1
      1) Should fail


  2 passing (7s)
  1 failing

  1) Main Suite 2 @parallel
       Subsuite 2-1
         Should fail:
     page.click: Timeout 5000ms exceeded.
```

worldomonation committed:
Do not export MediaHelper from @automattic/calypso-e2e module.

Remove dependency on mocha-steps in preparation for #52658 on trunk.

worldomonation committed:
Remove export of MediaHelper as per comments on the PR.

worldomonation commented:
Though this PR can probably land on its own, to be extra sure I will wait for Sergio's bail experiment to be merged to `trunk` prior to merging this one.
